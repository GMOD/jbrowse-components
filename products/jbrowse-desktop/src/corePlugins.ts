import Alignments from '@jbrowse/plugin-alignments'
import Bed from '@jbrowse/plugin-bed'
import BreakpointSplitView from '@jbrowse/plugin-breakpoint-split-view'
import CircularView from '@jbrowse/plugin-circular-view'
import Config from '@jbrowse/plugin-config'
import DataManagement from '@jbrowse/plugin-data-management'
import DotplotView from '@jbrowse/plugin-dotplot-view'
import Filtering from '@jbrowse/plugin-filtering'
import Gff3Tabix from '@jbrowse/plugin-gff3'
import LegacyJBrowse from '@jbrowse/plugin-legacy-jbrowse'
import LinearGenomeView from '@jbrowse/plugin-linear-genome-view'
import LinearComparativeView from '@jbrowse/plugin-linear-comparative-view'
import Lollipop from '@jbrowse/plugin-lollipop'
import Menus from '@jbrowse/plugin-menus'
import RDF from '@jbrowse/plugin-rdf'
import Sequence from '@jbrowse/plugin-sequence'
import SVG from '@jbrowse/plugin-svg'
import TrackHubRegistry from '@jbrowse/plugin-trackhub-registry'
import Variants from '@jbrowse/plugin-variants'
import Wiggle from '@jbrowse/plugin-wiggle'
import SpreadsheetViewPlugin from '@jbrowse/plugin-spreadsheet-view'
import SvInspectorPlugin from '@jbrowse/plugin-sv-inspector'
import HicPlugin from '@jbrowse/plugin-hic'

export default [
  SVG,
  LinearGenomeView,
  Alignments,
  Bed,
  CircularView,
  Config,
  DataManagement,
  DotplotView,
  Filtering,
  Gff3Tabix,
  LegacyJBrowse,
  LinearComparativeView,
  Lollipop,
  Menus,
  RDF,
  Sequence,
  TrackHubRegistry,
  Variants,
  Wiggle,
  SpreadsheetViewPlugin,
  SvInspectorPlugin,
  BreakpointSplitView,
  HicPlugin,
]
