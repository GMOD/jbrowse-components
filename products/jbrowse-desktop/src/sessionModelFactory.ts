/* eslint-disable @typescript-eslint/no-explicit-any */
import { AnyConfigurationModel } from '@jbrowse/core/configuration/configurationSchema'
import {
  readConfObject,
  isConfigurationModel,
} from '@jbrowse/core/configuration'
import {
  Region,
  NotificationLevel,
  TrackViewModel,
} from '@jbrowse/core/util/types'
import { getContainingView } from '@jbrowse/core/util'
import { observable } from 'mobx'
import {
  getMembers,
  getParent,
  getSnapshot,
  getType,
  IAnyStateTreeNode,
  isAlive,
  isModelType,
  isReferenceType,
  SnapshotIn,
  types,
  walk,
} from 'mobx-state-tree'
import PluginManager from '@jbrowse/core/PluginManager'

import SettingsIcon from '@material-ui/icons/Settings'
import CopyIcon from '@material-ui/icons/FileCopy'
import DeleteIcon from '@material-ui/icons/Delete'

declare interface ReferringNode {
  node: IAnyStateTreeNode
  key: string
}

export default function sessionModelFactory(pluginManager: PluginManager) {
  const minDrawerWidth = 128
  return types
    .model('JBrowseDesktopSessionModel', {
      name: types.identifier,
      margin: 0,
      drawerWidth: types.optional(
        types.refinement(types.integer, width => width >= minDrawerWidth),
        384,
      ),
      views: types.array(pluginManager.pluggableMstType('view', 'stateModel')),
      widgets: types.map(
        pluginManager.pluggableMstType('widget', 'stateModel'),
      ),
      activeWidgets: types.map(
        types.safeReference(
          pluginManager.pluggableMstType('widget', 'stateModel'),
        ),
      ),
      connectionInstances: types.map(
        types.array(pluginManager.pluggableMstType('connection', 'stateModel')),
      ),
    })
    .volatile((/* self */) => ({
      pluginManager,
      /**
       * this is the globally "selected" object. can be anything.
       * code that wants to deal with this should examine it to see what
       * kind of thing it is.
       */
      selection: undefined,
      /**
       * this is the current "task" that is being performed in the UI.
       * this is usually an object of the form
       * `{ taskName: "configure", target: thing_being_configured }`
       */
      task: undefined,
    }))
    .views(self => ({
      get rpcManager() {
        return getParent(self).jbrowse.rpcManager
      },
      get configuration() {
        return getParent(self).jbrowse.configuration
      },
      get assemblies() {
        return getParent(self).jbrowse.assemblies
      },
      get assemblyNames() {
        return getParent(self).jbrowse.assemblyNames
      },
      get tracks() {
        return getParent(self).jbrowse.tracks
      },
      get connections() {
        return getParent(self).jbrowse.connections
      },
      get savedSessions() {
        return getParent(self).jbrowse.savedSessions
      },
      get savedSessionNames() {
        return getParent(self).jbrowse.savedSessionNames
      },
      get history() {
        return getParent(self).history
      },
      get menus() {
        return getParent(self).menus
      },

      get assemblyManager() {
        return getParent(self).assemblyManager
      },
      get version() {
        return getParent(self).version
      },
      get renderProps() {
        return { theme: readConfObject(this.configuration, 'theme') }
      },
      get visibleWidget() {
        if (isAlive(self))
          // returns most recently added item in active widgets
          return Array.from(self.activeWidgets.values())[
            self.activeWidgets.size - 1
          ]
        return undefined
      },
      /**
       * See if any MST nodes currently have a types.reference to this object.
       * @param object - object
       * @returns An array where the first element is the node referring
       * to the object and the second element is they property name the node is
       * using to refer to the object
       */
      getReferring(object: IAnyStateTreeNode) {
        const refs: ReferringNode[] = []
        walk(getParent(self), node => {
          if (isModelType(getType(node))) {
            const members = getMembers(node)
            Object.entries(members.properties).forEach(([key, value]) => {
              // @ts-ignore
              if (isReferenceType(value) && node[key] === object) {
                refs.push({ node, key })
              }
            })
          }
        })
        return refs
      },
    }))
    .actions(self => ({
      makeConnection(
        configuration: AnyConfigurationModel,
        initialSnapshot = {},
      ) {
        const { type } = configuration
        if (!type) throw new Error('track configuration has no `type` listed')
        const name = readConfObject(configuration, 'name')
        const connectionType = pluginManager.getConnectionType(type)
        if (!connectionType) throw new Error(`unknown connection type ${type}`)
        const assemblyName = readConfObject(configuration, 'assemblyName')
        const connectionData = {
          ...initialSnapshot,
          name,
          type,
          configuration,
        }
        if (!self.connectionInstances.has(assemblyName))
          self.connectionInstances.set(assemblyName, [])
        const assemblyConnections = self.connectionInstances.get(assemblyName)
        if (!assemblyConnections)
          throw new Error(`assembly ${assemblyName} not found`)
        const length = assemblyConnections.push(connectionData)
        return assemblyConnections[length - 1]
      },

      prepareToBreakConnection(configuration: AnyConfigurationModel) {
        const callbacksToDereferenceTrack: Function[] = []
        const dereferenceTypeCount: Record<string, number> = {}
        const name = readConfObject(configuration, 'name')
        const assemblyName = readConfObject(configuration, 'assemblyName')
        const assemblyConnections =
          self.connectionInstances.get(assemblyName) || []
        const connection = assemblyConnections.find(c => c.name === name)
        if (connection) {
          connection.tracks.forEach((track: any) => {
            const referring = self.getReferring(track)
            this.removeReferring(
              referring,
              track,
              callbacksToDereferenceTrack,
              dereferenceTypeCount,
            )
          })
          const safelyBreakConnection = () => {
            callbacksToDereferenceTrack.forEach(cb => cb())
            this.breakConnection(configuration)
          }
          return [safelyBreakConnection, dereferenceTypeCount]
        }
        return undefined
      },

      breakConnection(configuration: AnyConfigurationModel) {
        const name = readConfObject(configuration, 'name')
        const assemblyName = readConfObject(configuration, 'assemblyName')
        const connectionInstances = self.connectionInstances.get(assemblyName)
        if (!connectionInstances)
          throw new Error(`connections for ${assemblyName} not found`)
        const connection = connectionInstances.find(c => c.name === name)
        connectionInstances.remove(connection)
      },

      deleteConnection(configuration: AnyConfigurationModel) {
        return getParent(self).jbrowse.deleteConnectionConf(configuration)
      },

      updateDrawerWidth(drawerWidth: number) {
        if (drawerWidth === self.drawerWidth) return self.drawerWidth
        let newDrawerWidth = drawerWidth
        if (newDrawerWidth < minDrawerWidth) newDrawerWidth = minDrawerWidth
        self.drawerWidth = newDrawerWidth
        return newDrawerWidth
      },

      resizeDrawer(distance: number) {
        const oldDrawerWidth = self.drawerWidth
        const newDrawerWidth = this.updateDrawerWidth(oldDrawerWidth - distance)
        const actualDistance = oldDrawerWidth - newDrawerWidth
        return actualDistance
      },

      addView(typeName: string, initialState = {}) {
        const typeDefinition = pluginManager.getElementType('view', typeName)
        if (!typeDefinition) throw new Error(`unknown view type ${typeName}`)

        const length = self.views.push({
          ...initialState,
          type: typeName,
        })
        return self.views[length - 1]
      },

      removeView(view: any) {
        for (const [id, widget] of self.activeWidgets) {
          if (
            id === 'hierarchicalTrackSelector' &&
            widget.view &&
            widget.view.id === view.id
          )
            this.hideWidget(widget)
        }
        self.views.remove(view)
      },

      addAssemblyConf(assemblyConf: any) {
        return getParent(self).jbrowse.addAssemblyConf(assemblyConf)
      },

      addTrackConf(trackConf: any) {
        return getParent(self).jbrowse.addTrackConf(trackConf)
      },

      hasWidget(widget: any) {
        return self.activeWidgets.has(widget.id)
      },

      removeReferring(
        referring: any,
        track: any,
        callbacks: Function[],
        dereferenceTypeCount: Record<string, number>,
      ) {
        referring.forEach(({ node }: ReferringNode) => {
          let dereferenced = false
          try {
            // If a view is referring to the track config, remove the track
            // from the view
            const type = 'open track(s)'
            const view = getContainingView(node) as TrackViewModel
            callbacks.push(() => view.hideTrack(track.trackId))
            dereferenced = true
            if (!dereferenceTypeCount[type]) {
              dereferenceTypeCount[type] = 0
            }
            dereferenceTypeCount[type] += 1
          } catch (err1) {
            // ignore
          }

          // @ts-ignore
          if (self.widgets.has(node.id)) {
            // If a configuration editor widget has the track config
            // open, close the widget
            const type = 'configuration editor widget(s)'
            callbacks.push(() => this.hideWidget(node))
            dereferenced = true
            if (!dereferenceTypeCount[type]) dereferenceTypeCount[type] = 0
            dereferenceTypeCount[type] += 1
          }
          if (!dereferenced)
            throw new Error(
              `Error when closing this connection, the following node is still referring to a track configuration: ${JSON.stringify(
                getSnapshot(node),
              )}`,
            )
        })
      },

      /**
       * See if any MST nodes currently have a types.reference to this object.
       * @param object - object
       * @returns An array where the first element is the node referring
       * to the object and the second element is they property name the node is
       * using to refer to the object
       */
      getReferring(object: IAnyStateTreeNode) {
        const refs: ReferringNode[] = []
        walk(getParent(self), node => {
          if (isModelType(getType(node))) {
            const members = getMembers(node)
            Object.entries(members.properties).forEach(([key, value]) => {
              // @ts-ignore
              if (isReferenceType(value) && node[key] === object) {
                refs.push({ node, key })
              }
            })
          }
        })
        return refs
      },
      deleteTrackConf(trackConf: AnyConfigurationModel) {
        const callbacksToDereferenceTrack: Function[] = []
        const dereferenceTypeCount: Record<string, number> = {}
        const referring = self.getReferring(trackConf)
        this.removeReferring(
          referring,
          trackConf,
          callbacksToDereferenceTrack,
          dereferenceTypeCount,
        )
        callbacksToDereferenceTrack.forEach(cb => cb())
        return getParent(self).jbrowse.deleteTrackConf(trackConf)
      },

      addConnectionConf(connectionConf: any) {
        return getParent(self).jbrowse.addConnectionConf(connectionConf)
      },

      addLinearGenomeViewOfAssembly(assemblyName: string, initialState = {}) {
        return this.addViewOfAssembly(
          'LinearGenomeView',
          assemblyName,
          initialState,
        )
      },

      addViewOfAssembly(
        viewType: any,
        assemblyName: string,
        initialState: any = {},
      ) {
        const assembly = self.assemblies.find(
          (s: AnyConfigurationModel) =>
            readConfObject(s, 'name') === assemblyName,
        )
        if (!assembly)
          throw new Error(
            `Could not add view of assembly "${assemblyName}", assembly name not found`,
          )
        initialState.displayRegionsFromAssemblyName = readConfObject(
          assembly,
          'name',
        )
        return this.addView(viewType, initialState)
      },

      addViewFromAnotherView(
        viewType: string,
        otherView: any,
        initialState: { displayedRegions?: Region[] } = {},
      ) {
        const state = { ...initialState }
        state.displayedRegions = getSnapshot(otherView.displayedRegions)
        return this.addView(viewType, state)
      },

      addWidget(
        typeName: string,
        id: string,
        initialState = {},
        configuration = { type: typeName },
      ) {
        const typeDefinition = pluginManager.getElementType('widget', typeName)
        if (!typeDefinition) throw new Error(`unknown widget type ${typeName}`)
        const data = {
          ...initialState,
          id,
          type: typeName,
          configuration,
        }
        self.widgets.set(id, data)
        return self.widgets.get(id)
      },

      showWidget(widget: any) {
        if (self.activeWidgets.has(widget.id))
          self.activeWidgets.delete(widget.id)
        self.activeWidgets.set(widget.id, widget)
      },

      hideWidget(widget: any) {
        self.activeWidgets.delete(widget.id)
      },

      hideAllWidgets() {
        self.activeWidgets.clear()
      },

      /**
       * set the global selection, i.e. the globally-selected object.
       * can be a feature, a view, just about anything
       * @param thing -
       */
      setSelection(thing: any) {
        self.selection = thing
      },

      /**
       * clears the global selection
       */
      clearSelection() {
        self.selection = undefined
      },

      /**
       * opens a configuration editor to configure the given thing,
       * and sets the current task to be configuring it
       * @param configuration -
       */
      editConfiguration(configuration: AnyConfigurationModel) {
        if (!isConfigurationModel(configuration)) {
          throw new Error(
            'must pass a configuration model to editConfiguration',
          )
        }
        const editor = this.addWidget(
          'ConfigurationEditorWidget',
          'configEditor',
          { target: configuration },
        )
        this.showWidget(editor)
      },
      editTrackConfiguration(configuration: AnyConfigurationModel) {
        this.editConfiguration(configuration)
      },

      clearConnections() {
        self.connectionInstances.clear()
      },

      addSavedSession(sessionSnapshot: SnapshotIn<typeof self>) {
        return getParent(self).jbrowse.addSavedSession(sessionSnapshot)
      },

      removeSavedSession(sessionSnapshot: any) {
        return getParent(self).jbrowse.removeSavedSession(sessionSnapshot)
      },

      renameCurrentSession(sessionName: string) {
        return getParent(self).renameCurrentSession(sessionName)
      },

      duplicateCurrentSession() {
        return getParent(self).duplicateCurrentSession()
      },

      activateSession(sessionName: any) {
        return getParent(self).activateSession(sessionName)
      },

      setDefaultSession() {
        return getParent(self).setDefaultSession()
      },
    }))
    .extend(() => {
      const snackbarMessages = observable.array()

      return {
        views: {
          get snackbarMessages() {
            return snackbarMessages
          },
        },
        actions: {
          notify(message: string, level?: NotificationLevel) {
            return this.pushSnackbarMessage(message, level)
          },

          pushSnackbarMessage(message: string, level?: NotificationLevel) {
            return snackbarMessages.push([message, level])
          },

          popSnackbarMessage() {
            return snackbarMessages.pop()
          },
        },
      }
    })
    .views(self => ({
      getTrackActionMenuItems(config: any) {
        const session = self
        return [
          {
            label: 'Settings',
            onClick: () => {
              session.editConfiguration(config)
            },
            icon: SettingsIcon,
          },
          {
            label: 'Delete track',
            onClick: () => {
              session.deleteTrackConf(config)
            },
            icon: DeleteIcon,
          },
          {
            label: 'Copy track',
            onClick: () => {
              const trackSnapshot = JSON.parse(
                JSON.stringify(getSnapshot(config)),
              )
              const now = Date.now()
              trackSnapshot.trackId += `-${now}`
              trackSnapshot.displays.forEach(
                (display: { displayId: string }) => {
                  display.displayId += `-${now}`
                },
              )
              trackSnapshot.name += ' (copy)'
              trackSnapshot.category = undefined
              session.addTrackConf(trackSnapshot)
            },
            icon: CopyIcon,
          },
        ]
      },
    }))
}

export type SessionStateModel = ReturnType<typeof sessionModelFactory>

// a track is a combination of an assembly and a renderer, along with some conditions
// specifying in which contexts it is available (which assemblies, which views, etc)
