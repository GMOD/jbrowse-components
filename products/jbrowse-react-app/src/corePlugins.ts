import Alignments from '@jbrowse/plugin-alignments'
import Arc from '@jbrowse/plugin-arc'
import Authentication from '@jbrowse/plugin-authentication'
import Bed from '@jbrowse/plugin-bed'
import BreakpointSplitView from '@jbrowse/plugin-breakpoint-split-view'
import CircularView from '@jbrowse/plugin-circular-view'
import ComparativeAdapters from '@jbrowse/plugin-comparative-adapters'
import Config from '@jbrowse/plugin-config'
import DataManagement from '@jbrowse/plugin-data-management'
import DotplotView from '@jbrowse/plugin-dotplot-view'
import GCContent from '@jbrowse/plugin-gccontent'
import Gff3 from '@jbrowse/plugin-gff3'
import GridBookmarkPlugin from '@jbrowse/plugin-grid-bookmark'
import GtfPlugin from '@jbrowse/plugin-gtf'
import HicPlugin from '@jbrowse/plugin-hic'
import LegacyJBrowse from '@jbrowse/plugin-legacy-jbrowse'
import LinearComparativeView from '@jbrowse/plugin-linear-comparative-view'
import LinearGenomeView from '@jbrowse/plugin-linear-genome-view'
import Lollipop from '@jbrowse/plugin-lollipop'
import Menus from '@jbrowse/plugin-menus'
import RDF from '@jbrowse/plugin-rdf'
import Sequence from '@jbrowse/plugin-sequence'
import SpreadsheetViewPlugin from '@jbrowse/plugin-spreadsheet-view'
import SvInspectorPlugin from '@jbrowse/plugin-sv-inspector'
import SVG from '@jbrowse/plugin-svg'
import TrixPlugin from '@jbrowse/plugin-trix'
import Variants from '@jbrowse/plugin-variants'
import Wiggle from '@jbrowse/plugin-wiggle'

const corePlugins = [
  SVG,
  LinearGenomeView,
  Alignments,
  Authentication,
  Bed,
  CircularView,
  Config,
  DataManagement,
  DotplotView,
  GtfPlugin,
  Gff3,
  LegacyJBrowse,
  LinearComparativeView,
  Lollipop,
  Menus,
  RDF,
  Sequence,
  Variants,
  Wiggle,
  GCContent,
  SpreadsheetViewPlugin,
  SvInspectorPlugin,
  BreakpointSplitView,
  HicPlugin,
  TrixPlugin,
  GridBookmarkPlugin,
  ComparativeAdapters,
  Arc,
]

export default corePlugins
