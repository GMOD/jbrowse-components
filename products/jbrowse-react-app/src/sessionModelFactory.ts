import oldSessionModelFactory from './sessionModel'

/** @deprecated moved to ./sessionModel */
const sessionModelFactory = oldSessionModelFactory
export default sessionModelFactory
