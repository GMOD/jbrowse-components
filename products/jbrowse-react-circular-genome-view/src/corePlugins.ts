import Config from '@jbrowse/plugin-config'
import DataManagement from '@jbrowse/plugin-data-management'
import CircularGenomeView from '@jbrowse/plugin-circular-view'
import Sequence from '@jbrowse/plugin-sequence'
import Variants from '@jbrowse/plugin-variants'
import Wiggle from '@jbrowse/plugin-wiggle'

const corePlugins = [
  Config,
  DataManagement,
  CircularGenomeView,
  Sequence,
  Variants,
  Wiggle,
]

export default corePlugins
