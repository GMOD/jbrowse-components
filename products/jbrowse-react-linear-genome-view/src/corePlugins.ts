import Alignments from '@jbrowse/plugin-alignments'
import BED from '@jbrowse/plugin-bed'
import Config from '@jbrowse/plugin-config'
import DataManagement from '@jbrowse/plugin-data-management'
import GFF3 from '@jbrowse/plugin-gff3'
import LinearGenomeView from '@jbrowse/plugin-linear-genome-view'
import Sequence from '@jbrowse/plugin-sequence'
import SVG from '@jbrowse/plugin-svg'
import Variants from '@jbrowse/plugin-variants'
import Wiggle from '@jbrowse/plugin-wiggle'

export default [
  Alignments,
  BED,
  Config,
  DataManagement,
  GFF3,
  LinearGenomeView,
  Sequence,
  SVG,
  Variants,
  Wiggle,
]
