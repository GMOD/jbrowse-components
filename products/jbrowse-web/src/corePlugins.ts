import Alignments from '@jbrowse/plugin-alignments'
import Authentication from '@jbrowse/plugin-authentication'
import Bed from '@jbrowse/plugin-bed'
import BreakpointSplitView from '@jbrowse/plugin-breakpoint-split-view'
import CircularView from '@jbrowse/plugin-circular-view'
import Config from '@jbrowse/plugin-config'
import DataManagement from '@jbrowse/plugin-data-management'
import DotplotView from '@jbrowse/plugin-dotplot-view'
import Gff3 from '@jbrowse/plugin-gff3'
import LegacyJBrowse from '@jbrowse/plugin-legacy-jbrowse'
import LinearGenomeView from '@jbrowse/plugin-linear-genome-view'
import LinearComparativeView from '@jbrowse/plugin-linear-comparative-view'
import Lollipop from '@jbrowse/plugin-lollipop'
import Menus from '@jbrowse/plugin-menus'
import RDF from '@jbrowse/plugin-rdf'
import Sequence from '@jbrowse/plugin-sequence'
import SVG from '@jbrowse/plugin-svg'
import TrackHubRegistry from '@jbrowse/plugin-trackhub-registry'
import Variants from '@jbrowse/plugin-variants'
import Wiggle from '@jbrowse/plugin-wiggle'
import SpreadsheetViewPlugin from '@jbrowse/plugin-spreadsheet-view'
import SvInspectorPlugin from '@jbrowse/plugin-sv-inspector'
import HicPlugin from '@jbrowse/plugin-hic'
import FlashmapPlugin from '@jbrowse/plugin-flashmap'
import TrixPlugin from '@jbrowse/plugin-trix'
import GridBookmarkPlugin from '@jbrowse/plugin-grid-bookmark'

const corePlugins = [
  SVG,
  LinearGenomeView,
  Alignments,
  Authentication,
  Bed,
  CircularView,
  Config,
  DataManagement,
  DotplotView,
  Gff3,
  LegacyJBrowse,
  LinearComparativeView,
  Lollipop,
  Menus,
  RDF,
  Sequence,
  TrackHubRegistry,
  Variants,
  Wiggle,
  SpreadsheetViewPlugin,
  SvInspectorPlugin,
  BreakpointSplitView,
  HicPlugin,
  FlashmapPlugin,
  TrixPlugin,
  GridBookmarkPlugin,
]

export default corePlugins
