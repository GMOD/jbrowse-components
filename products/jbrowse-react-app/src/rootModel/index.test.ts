// we use mainthread rpc so we mock the makeWorkerInstance to an empty file
import PluginManager from '@jbrowse/core/PluginManager'
import { getSnapshot } from 'mobx-state-tree'
import corePlugins from '../corePlugins'
import rootModelFactory from './'
import sessionModelFactory from '../sessionModel'

jest.mock('../makeWorkerInstance', () => () => {})

function getRootModel() {
  const pluginManager = new PluginManager(corePlugins.map(P => new P()))
  pluginManager.createPluggableElements()
  pluginManager.configure()
  return rootModelFactory({ pluginManager, sessionModelFactory })
}
describe('Root MST model', () => {
  afterEach(() => {
    localStorage.clear()
    sessionStorage.clear()
  })

  it('creates with defaults', () => {
    const root = getRootModel().create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
    })
    expect(root.session).toBeUndefined()
    root.setDefaultSession()
    expect(root.session).toBeTruthy()
    expect(root.jbrowse.assemblies.length).toBe(0)
    expect(getSnapshot(root.jbrowse.configuration)).toMatchSnapshot()
  })

  it('creates with a minimal session', () => {
    const root = getRootModel().create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
      session: { name: 'testSession' },
    })
    expect(root.session).toBeTruthy()
  })

  it('activates a session snapshot', () => {
    const session = { name: 'testSession' }
    localStorage.setItem('localSaved-123', JSON.stringify({ session }))
    Storage.prototype.getItem = jest.fn(
      () => `{"session": {"name": "testSession"}}`,
    )
    const root = getRootModel().create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
    })
    expect(root.session).toBeUndefined()
    root.setSession(session)
    expect(root.session).toBeTruthy()
  })

  it('adds track and connection configs to an assembly', () => {
    const root = getRootModel().create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
        assemblies: [
          {
            name: 'assembly1',
            aliases: ['assemblyA'],
            sequence: {
              trackId: 'sequenceConfigId',
              type: 'ReferenceSequenceTrack',
              adapter: {
                type: 'FromConfigSequenceAdapter',
                adapterId: 'sequenceConfigAdapterId',
                features: [
                  {
                    refName: 'ctgA',
                    uniqueId: 'firstId',
                    start: 0,
                    end: 10,
                    seq: 'cattgttgcg',
                  },
                ],
              },
            },
          },
        ],
      },
    })
    expect(root.jbrowse.assemblies.length).toBe(1)
    expect(getSnapshot(root.jbrowse.assemblies[0])).toMatchSnapshot()
    const newTrackConf = root.jbrowse.addTrackConf({
      type: 'FeatureTrack',
      trackId: 'trackId0',
    })
    expect(getSnapshot(newTrackConf)).toMatchSnapshot()
    expect(root.jbrowse.tracks.length).toBe(1)
    const newConnectionConf = root.jbrowse.addConnectionConf({
      type: 'JBrowse1Connection',
      connectionId: 'connectionId0',
    })
    expect(getSnapshot(newConnectionConf)).toMatchSnapshot()
    expect(root.jbrowse.connections.length).toBe(1)
  })

  it('throws if session is invalid', () => {
    expect(() =>
      getRootModel().create({
        jbrowse: {
          configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
        },
        session: {},
      }),
    ).toThrow()
  })

  it('throws if session snapshot is invalid', () => {
    const root = getRootModel().create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
    })
    expect(() => {
      root.setSession({})
    }).toThrow()
  })

  it('adds menus', () => {
    const root = getRootModel().create({
      jbrowse: {
        configuration: { rpc: { defaultDriver: 'MainThreadRpcDriver' } },
      },
    })
    expect(root.menus).toMatchSnapshot()
    root.appendMenu('Third Menu')
    root.insertMenu('Second Menu', -1)
    root.appendToMenu('Second Menu', {
      label: 'Second Menu Item',
      onClick: () => {},
    })
    root.insertInMenu(
      'Second Menu',
      {
        label: 'First Menu Item',
        onClick: () => {},
      },
      -1,
    )
    root.appendToSubMenu(['Second Menu', 'First Sub Menu'], {
      label: 'Second Sub Menu Item',
      onClick: () => {},
    })
    root.insertInSubMenu(
      ['Second Menu', 'First Sub Menu'],
      {
        label: 'First Sub Menu Item',
        onClick: () => {},
      },
      -1,
    )
    expect(root.menus).toMatchSnapshot()
    expect(() => {
      root.appendToSubMenu(['Second Menu', 'First Menu Item'], {
        label: 'First Sub Menu Item',
        onClick: () => {},
      })
    }).toThrow(/is not a subMenu/)
  })
})
