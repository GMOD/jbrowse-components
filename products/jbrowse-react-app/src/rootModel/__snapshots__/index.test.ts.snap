// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Root MST model adds menus 1`] = `
[
  {
    "label": "File",
    "menuItems": [
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "New session",
        "onClick": [Function],
      },
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "Import session…",
        "onClick": [Function],
      },
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "Export session",
        "onClick": [Function],
      },
      {
        "type": "divider",
      },
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "Open track...",
        "onClick": [Function],
      },
      {
        "icon": [Function],
        "label": "Open connection...",
        "onClick": [Function],
      },
    ],
  },
  {
    "label": "Add",
    "menuItems": [],
  },
  {
    "label": "Tools",
    "menuItems": [],
  },
]
`;

exports[`Root MST model adds menus 2`] = `
[
  {
    "label": "File",
    "menuItems": [
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "New session",
        "onClick": [Function],
      },
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "Import session…",
        "onClick": [Function],
      },
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "Export session",
        "onClick": [Function],
      },
      {
        "type": "divider",
      },
      {
        "icon": {
          "$$typeof": Symbol(react.memo),
          "compare": null,
          "type": {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "label": "Open track...",
        "onClick": [Function],
      },
      {
        "icon": [Function],
        "label": "Open connection...",
        "onClick": [Function],
      },
    ],
  },
  {
    "label": "Add",
    "menuItems": [],
  },
  {
    "label": "Tools",
    "menuItems": [],
  },
  {
    "label": "Second Menu",
    "menuItems": [
      {
        "label": "First Menu Item",
        "onClick": [Function],
      },
      {
        "label": "Second Menu Item",
        "onClick": [Function],
      },
      {
        "label": "First Sub Menu",
        "subMenu": [
          {
            "label": "First Sub Menu Item",
            "onClick": [Function],
          },
          {
            "label": "Second Sub Menu Item",
            "onClick": [Function],
          },
        ],
      },
    ],
  },
  {
    "label": "Third Menu",
    "menuItems": [],
  },
]
`;

exports[`Root MST model adds track and connection configs to an assembly 1`] = `
{
  "aliases": [
    "assemblyA",
  ],
  "name": "assembly1",
  "sequence": {
    "adapter": {
      "adapterId": "sequenceConfigAdapterId",
      "features": [
        {
          "end": 10,
          "refName": "ctgA",
          "seq": "cattgttgcg",
          "start": 0,
          "uniqueId": "firstId",
        },
      ],
      "type": "FromConfigSequenceAdapter",
    },
    "displays": [
      {
        "displayId": "sequenceConfigId-LinearReferenceSequenceDisplay",
        "type": "LinearReferenceSequenceDisplay",
      },
      {
        "displayId": "sequenceConfigId-LinearGCContentDisplay",
        "type": "LinearGCContentDisplay",
      },
    ],
    "trackId": "sequenceConfigId",
    "type": "ReferenceSequenceTrack",
  },
}
`;

exports[`Root MST model adds track and connection configs to an assembly 2`] = `
{
  "displays": [
    {
      "displayId": "trackId0-LinearBasicDisplay",
      "type": "LinearBasicDisplay",
    },
    {
      "displayId": "trackId0-LinearArcDisplay",
      "type": "LinearArcDisplay",
    },
  ],
  "trackId": "trackId0",
  "type": "FeatureTrack",
}
`;

exports[`Root MST model adds track and connection configs to an assembly 3`] = `
{
  "connectionId": "connectionId0",
  "dataDirLocation": {
    "internetAccountId": undefined,
    "internetAccountPreAuthorization": undefined,
    "locationType": "UriLocation",
    "uri": "http://mysite.com/jbrowse/data/",
  },
  "type": "JBrowse1Connection",
}
`;

exports[`Root MST model creates with defaults 1`] = `{}`;
